proposals/data-classes.md

```diff
@@ -0,0 +1,372 @@
+`LoginResource` class that people write today.
+
+The LoginResource class would now be defined as
+
+```C#
+public data class LoginResource
+{
+    public string Username { get; set; }
+    public string Password { get; set; }
+    public bool RememberMe { get; set; }
+}
+```
+
+and the use would be identical:
+
+```C#
+var x = new LoginResource {
+    Username = "andy",
+    Password = password
+}
+```
+
+However, the generated class code would look like:
+
+```C#
+public class LoginResource : IEquatable<LoginResource>
+{
+    public string Username { get; set; }
+    public string Password { get; set; }
+    public bool RememberMe { get; set; }
+
+    public override bool Equals(object obj)
+        => obj is LoginResource resource && Equals(resource);
+
+    public bool Equals(LoginResource other)
+    {
+        return other != null &&
+               EqualityContractOrigin == other.EqualityContractOrigin &&
+               Username == other.Username &&
+               Password == other.Password &&
+               RememberMe == other.RememberMe;
+    }
+
+    protected virtual Type EqualityContractOrigin => typeof(LoginResource);
+
+    public override int GetHashCode()
+    {
+        unchecked
+        {
+            return EqualityComparer<string>.Default.GetHashCode(Username) +
+                EqualityComparer<string>.Default.GetHashCode(Password) +
+                RememberMe.GetHashCode();
+        }
+    }
+
+    public override string ToString()
+    {
+        return $"{{{nameof(Username)} = {Username}, {nameof(Password)} = {Password}, {nameof(RememberMe)} = {RememberMe}}}";
+    }
+
+    public static bool operator ==(LoginResource resource1, LoginResource resource2)
+    {
+        return EqualityComparer<LoginResource>.Default.Equals(resource1, resource2);
+    }
+
+    public static bool operator !=(LoginResource resource1, LoginResource resource2)
+    {
+        return !(resource1 == resource2);
+    }
+}
+```
+
+### Equality
+
+First, the generation of equality support. Data members are only public
+fields and auto-properties. This allows data classes to have private
+implementation details without giving up simple equality semantics. There are
+a few places this could be problematic. For instance, only auto-properties
+are considered data members by default, but it's not uncommon to have some
+simple validation included in a property getter that does not meaningfully
+change the semantics, e.g.
+
+```C#
+{
+    ...
+    private int _field;
+    public int Field
+    {
+        get
+        {
+            Debug.Assert(_field >= 0);
+            return _field;
+        }
+        set { ... }
+    }
+}
+```
+
+To support these cases and provide an easy escape hatch, I propose a
+new attribute, `DataMemberAttribute` with a boolean flag argument on the
+constructor. This allows users to override the normal behavior and include
+or exclude extra members in equality. The previous example would now read:
```

Comment:
this seems odd to me.  Given that it's a Data-Class or Data-Struct, my bias is that the public props/fields should be part of the data-ness.  It seems like if someone does not want that, the attribute should be to *opt-out*, instead of needing to opt-in.

Reply:
*auto*-properties are part of data-ness, computed properties may not be. That's especially true if you have, say, a half dozen computed properties all reading from single piece of data that you could compare directly for equality. For instance, flags enums to boolean properties.

Reply:
> auto-properties are part of data-ness, computed properties may not be.

Agreed.  but my point is simply: when something "may not be", what side of hte fence do you decide they're on.  My argument is we shoudl just assume they *are* part of the data-ness.  And if you don't want that, you opt-out.

Note: i also think opt-out is a good idea because maybe i want an auto-prop and i do not want that auto-prop to be part of the data-ness.  Now, i'd have a consistent way for auto-props and computed-props to say that.  i.e.:

```c#
[NotData]
internal int LogLevel { get; set; }
```

I want this to be an auto-prop because i really don't need to compute anything.  But i don't want it participating in the data-ness.

Reply:
@333fred I think in that case, the best thing would be to just provide your own override of GetHashCode/Equals.  i.e. you know better.  So you can give the fastest impl.

Reply:
@333fred That's what the proposal is talking about. You put `[DataMember(true)]` on the private field and it's considered in equality. The public computed properties are not.

I think @CyrusNajmabadi's approach is too heavy-handed.

Reply:
:D

Note: in fred's example, he jsut wants the private field used for equals/gethashcode.  Is there then a way to opt-out of something being a data member?  i.e. even if you have an auto-prop, can you put ```[DataMember(false)]``` on it? 

If so, i think i'm happy.  Basically, the language would have rules about if an unadorned member was a data-member or not (i.e. auto-props are, blah blah blah are not).  But you can always put on ```DataMember(true/false)``` on anything to explicitly opt-in/out.

Does that make sense ot you Andy?

Reply:
@CyrusNajmabadi Yeah, the boolean flag in the constructor was supposed to work both ways. To both opt-in and opt-out.